src/modules/flow/twitter/twitter.c

```diff
@@ -0,0 +1,617 @@
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "sol-flow/twitter.h"
+#include "sol-flow.h"
+#include "sol-http-client.h"
+#include "sol-http.h"
+#include "sol-json.h"
+#include "sol-mainloop.h"
+#include "sol-message-digest.h"
+#include "sol-random.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+#include "sol_config.h"
+
+#define BASE_POST_URL "https://api.twitter.com/1.1/statuses/update.json"
+#define BASE_TIMELINE_URL "https://api.twitter.com/1.1/statuses/home_timeline.json"
+
+struct twitter_data {
+    struct sol_ptr_vector pending_conns;
+    char *consumer_key, *consumer_secret,
+    *token, *token_secret;
+    char *escaped_post_url, *escaped_get_url;
+};
+
+struct callback_data {
+    struct sol_flow_node *node;
+    char *nonce, *timestamp, *status, *key;
+    struct sol_http_client_connection *(*cb)(struct callback_data *cb_data,
+        const char *authorization_header);
+};
+
+static char *
+generate_nonce(void)
+{
+    int r;
+    ssize_t size;
+    char *nonce = NULL;
+    struct sol_random *engine;
+    struct sol_str_slice slice;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+
+    engine = sol_random_new(SOL_RANDOM_DEFAULT, 0);
+    SOL_NULL_CHECK(engine, NULL);
+
+    size = sol_random_fill_buffer(engine, &buffer, 16);
+    sol_random_del(engine);
+    SOL_INT_CHECK_GOTO(size, < 16, end);
+
+    slice = SOL_STR_SLICE_STR(buffer.data, buffer.used);
+    size = sol_util_base16_calculate_encoded_len(slice);
+    SOL_INT_CHECK_GOTO(size, < 0, end);
+
+    nonce = calloc(1, size + 1);
+    SOL_NULL_CHECK_GOTO(nonce, end);
+
+    r = sol_util_base16_encode(nonce, size, slice, false);
+    if (r < 0) {
+        SOL_WRN("Could not enconde the oauth_nonce");
+        free(nonce);
+        nonce = NULL;
+    } else {
+        nonce[size] = '\0';
```

Comment:
Is this needed since you allocated `nonce` with `calloc`? But maybe allocating with `malloc` is better, then leaving this `\0` set as is.